app/Models/Concerns/HasMemberships.php

```diff
@@ -0,0 +1,148 @@
+use Carbon\Carbon;
+use Illuminate\Database\Eloquent\Relations\BelongsToMany;
+
+trait HasMemberships
+{
+    public function memberships(): BelongsToMany
+    {
+        return $this->membershipsRelationship()
+            ->wherePivot('ended_at', null)
+            ->orderByPriority();
+    }
+
+    public function membershipHistory(): BelongsToMany
+    {
+        return $this->membershipsRelationship()
+            ->orderBy((new MembershipPivot())->getTable().'.started_at', 'desc');
+    }
+
+    public function primaryMembership(): ?Membership
+    {
+        return $this->memberships()
+            ->primary()
+            ->first();
+    }
+
+    public function secondaryMemberships(): BelongsToMany
+    {
+        return $this->memberships()
+            ->secondary()
+            ->orderByPriority();
+    }
+
+    public function updatePrimaryMembership(string $division, string $region): bool
+    {
+        $matchingMembership = resolve(Membership::class)->findPrimaryByVATSIMLocality($division, $region);
+
+        if (! $matchingMembership) {
+            return false;
+        }
+
+        // Check if the user already has this membership, and the division/region combination is the same
+        if ($this->hasMembership($matchingMembership) && ($this->primaryMembership()->pivot->division == $division && $this->primaryMembership()->pivot->region == $region)) {
+            return false;
+        }
+
+        // Check we can have secondary memberships
+        if (! $matchingMembership->can_have_secondaries) {
+            $this->removeSecondaryMemberships();
+        }
+
+        // End Previous Primary Membership
+        if ($currentMembership = $this->primaryMembership()) {
+            $this->removeMembership($currentMembership);
+        }
+
+        $this->memberships()->attach($matchingMembership, [
+            'division' => $division,
+            'region' => $region,
+            'started_at' => Carbon::now(),
+        ]);
+
+        return true;
+    }
+
+    public function addSecondaryMembership(Membership $membership): bool
+    {
+        throw_if($membership->primary, new MembershipNotSecondaryException());
+
+        // Check we can have secondary memberships
+        if ($this->primaryMembership() && ! $this->primaryMembership()->can_have_secondaries) {
+            throw new PrimaryMembershipDoesntAllowSecondaryException();
+        }
+
+        // Check if the user already has this membership
+        if ($this->hasMembership($membership)) {
+            return false;
+        }
+
+        $this->memberships()->attach($membership);
+
+        return true;
+    }
+
+    public function removeSecondaryMemberships(): void
+    {
+        $this->secondaryMemberships->each(function ($membership) {
+            $this->removeMembership($membership);
+        });
+    }
+
+    /**
+     * @param Membership|int|string $membership
+     * @return bool
+     */
+    public function hasMembership($membership): bool
```

Comment:
Would it not be easier to enforce the model? This is a lot of code to test. If not, you could extract to methods like:

- hasMembershipByIdentifier
- hasMembershipById
- hasMembershipByModel

Reply:
Imo its better to have these options, because this way you don't have to make unnecessary database calls. The reason I combine it in one method like this is because this signature of accepting multiple ways to identify an object is pretty common in most Laravel functions.

There are only 3 paths through the method, and as such the test for it is only 3 lines? Extracting the functionality to separate methods seems overkill, and would remove the point of having the flexibility for the input identifier for the model